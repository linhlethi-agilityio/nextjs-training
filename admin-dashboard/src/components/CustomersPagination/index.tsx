// Api
import { getCustomers } from '@/api';

// Constants
import { DEFAULT_LIMIT } from '@/constants';

// Components
import { Pagination } from '@/components';

interface ProductPaginationProps {
  limit?: number;
  query?: string;
}

const CustomersPagination = async ({
  limit = DEFAULT_LIMIT,
  query = '',
}: ProductPaginationProps) => {
  const { data: customers = [] } = await getCustomers({ query });

  const totalPage = Math.ceil(customers?.length / limit);

  return customers.length !== 0 && <Pagination totalPage={totalPage} />;
};

export default CustomersPagination;
