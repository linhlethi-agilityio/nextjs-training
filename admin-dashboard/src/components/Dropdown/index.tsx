'use client';

import { ChangeEvent, memo } from 'react';
import { Select, SelectProps } from '@chakra-ui/react';
import isEqual from 'react-fast-compare';

export interface Option {
  value: string;
  label: string;
}

interface DropdownProps extends SelectProps {
  options: Option[];
  color?: string;
  placeholder?: string;
  onChangeValue: (value: string) => void;
}

const DropdownUI = ({
  options,
  placeholder = '',
  color = 'textDefault',
  width = 108,
  onChangeValue,
  ...rest
}: DropdownProps) => {
  /**
   * Function handle select dropdown
   */
  const handleSelect = (e: ChangeEvent<HTMLSelectElement>) => {
    const newValue = e.target.value;

    onChangeValue(newValue);
  };

  return (
    <Select
      placeholder={placeholder}
      color={color}
      aria-label="Select"
      onChange={handleSelect}
      lineHeight={2}
      w={width}
      {...rest}
    >
      {options.map((option) => {
        const { value = '', label = '' } = option;

        return (
          <option key={value} value={value}>
            {label}
          </option>
        );
      })}
    </Select>
  );
};

export const Dropdown = memo(DropdownUI, isEqual);
