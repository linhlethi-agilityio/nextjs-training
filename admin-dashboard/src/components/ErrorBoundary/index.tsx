'use client';

import { useEffect } from 'react';
import { Button, Flex, Heading } from '@chakra-ui/react';

interface ErrorProps {
  error: Error & { digest?: string };
  reset: () => void;
}

const ErrorBoundary = ({ error, reset }: ErrorProps) => {
  // Attempt to recover by trying to re-render the segment
  const handleReset = () => {
    reset();
  };

  useEffect(() => {
    // Log the error to an error reporting service
    console.error(error);
  }, [error]);

  return (
    <Flex flexDir="column" alignItems="center" gap="50px" mt={300}>
      <Heading>Something went wrong!</Heading>
      <Button onClick={handleReset}>Try again</Button>
    </Flex>
  );
};

export default ErrorBoundary;
