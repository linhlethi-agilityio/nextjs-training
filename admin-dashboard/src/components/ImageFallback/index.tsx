'use client';

import Image, { ImageProps } from 'next/image';
import { memo, useState } from 'react';

interface ImageFallbackProps extends ImageProps {
  src: string;
  fallbackSrc: string;
  alt: string;
}

const ImageFallback = ({
  src,
  fallbackSrc,
  alt,
  ...rest
}: ImageFallbackProps) => {
  const [imgSrc, setImgSrc] = useState(src);

  return (
    <Image
      src={imgSrc}
      blurDataURL={fallbackSrc}
      placeholder="blur"
      onLoadingComplete={(result) => {
        if (result.naturalWidth === 0) {
          // Broken image
          setImgSrc(fallbackSrc);
        }
      }}
      onError={() => {
        setImgSrc(fallbackSrc);
      }}
      alt={alt}
      {...rest}
    />
  );
};

export default memo(ImageFallback);
