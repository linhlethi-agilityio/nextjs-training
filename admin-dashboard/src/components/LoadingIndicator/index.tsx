import { Box, Spinner as SpinnerChakraUI } from '@chakra-ui/react';

const LoadingIndicator = (): JSX.Element => {
  return (
    <Box
      h="100vh"
      display="flex"
      alignItems="center"
      textAlign="center"
      justifyContent="center"
    >
      <SpinnerChakraUI
        size="lg"
        thickness="4px"
        speed="0.65s"
        color="brand.500"
      />
    </Box>
  );
};

export default LoadingIndicator;
