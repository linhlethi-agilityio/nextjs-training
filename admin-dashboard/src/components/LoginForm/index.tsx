'use client';

import { useMemo, KeyboardEvent, useState, memo, useTransition } from 'react';
import { Controller, useForm } from 'react-hook-form';
import { useRouter } from 'next/navigation';
import {
  Box,
  Button,
  FormControl,
  FormHelperText,
  Input,
  InputGroup,
  InputLeftElement,
  InputRightElement,
  useToast,
} from '@chakra-ui/react';

// Constants
import { ERROR_MESSAGES, ROUTES, SUCCESS_MESSAGES } from '@/constants';

// Utils
import {
  clearErrorOnChange,
  isEnableSubmitButton,
  isValidEmail,
} from '@/utils';

// Icons
import { EmailIcon, HideEyeIcon, PasswordIcon, ShowEyeIcon } from '@/icons';

interface LoginFormData {
  email: string;
  password: string;
}

const REQUIRED_FIELDS = ['email', 'password'];

interface LoginForm {
  onSubmit: (data: LoginFormData) => Promise<void | string>;
}

const LoginForm = ({ onSubmit }: LoginForm) => {
  const [showPassword, setShowPassword] = useState<boolean>(false);

  const toast = useToast();
  const [isPending, startTransition] = useTransition();
  const router = useRouter();

  const {
    control,
    formState: { dirtyFields, errors },
    clearErrors,
    handleSubmit,
  } = useForm<LoginFormData>({
    mode: 'onBlur',
    reValidateMode: 'onBlur',
    defaultValues: {
      email: '',
      password: '',
    },
  });

  // Checking to disable/enable submit button
  const dirtyItems = Object.keys(dirtyFields);

  const enableSubmit: boolean = useMemo(
    () => isEnableSubmitButton(REQUIRED_FIELDS, dirtyItems, errors),
    [dirtyItems, errors],
  );
  const isDisableSubmit = !enableSubmit;

  /**
   * Func handle login
   */
  const handleLogin = async (formData: LoginFormData) => {
    startTransition(async () => {
      const response = await onSubmit(formData);
      console.log('');

      if (typeof response === 'string') {
        toast({
          title: response,
          status: 'error',
        });
      } else {
        toast({
          title: SUCCESS_MESSAGES.LOGIN_SUCCESS,
          status: 'success',
        });
        router.push(ROUTES.PRODUCT);
      }
    });
  };

  const handleClickEyePassword = () => setShowPassword(!showPassword);

  const handleOnKeyDown = (event: KeyboardEvent<HTMLDivElement>) => {
    if (event.key === 'Enter') {
      handleSubmit(handleLogin)();
    }
  };

  return (
    <FormControl mt={50} onKeyDown={handleOnKeyDown} maxW={500}>
      <Controller
        name="email"
        control={control}
        rules={{
          required: ERROR_MESSAGES.FIELD_REQUIRED,
          validate: (email) =>
            isValidEmail(email) || ERROR_MESSAGES.INVALID_EMAIL,
        }}
        render={({
          field: { name, onChange, ...rest },
          fieldState: { error },
        }) => (
          <Box
            marginBottom={error?.message ? 0 : 26}
            _placeholder={{ color: 'textPlaceholder' }}
            ml={3}
          >
            <InputGroup>
              <InputLeftElement pointerEvents="none">
                <EmailIcon color="textPlaceholder" />
              </InputLeftElement>
              <Input
                id="email"
                placeholder="Your email"
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
            </InputGroup>

            {error?.message && (
              <FormHelperText
                color="red.400"
                fontSize="xs"
                mb={1}
                textAlign="left"
              >
                {error.message}
              </FormHelperText>
            )}
          </Box>
        )}
      />

      {/* Password */}
      <Controller
        name="password"
        control={control}
        rules={{
          required: ERROR_MESSAGES.FIELD_REQUIRED,
        }}
        render={({
          field: { name, onChange, ...rest },
          fieldState: { error },
        }) => (
          <Box
            marginBottom={error?.message ? 0 : 26}
            ml={3}
            _placeholder={{ color: 'textPlaceholder' }}
          >
            <InputGroup>
              <InputLeftElement pointerEvents="none">
                <PasswordIcon />
              </InputLeftElement>
              <InputRightElement>
                <Button variant="icon" onClick={handleClickEyePassword}>
                  {showPassword ? <ShowEyeIcon /> : <HideEyeIcon />}
                </Button>
              </InputRightElement>
              <Input
                w="full"
                id="password"
                placeholder="Password"
                type={showPassword ? 'text' : 'password'}
                isInvalid={!!error?.message}
                onChange={(e) => {
                  onChange(e.target.value);

                  // Clear error message on change
                  clearErrorOnChange(name, errors, clearErrors);
                }}
                {...rest}
              />
            </InputGroup>

            {error?.message && (
              <FormHelperText
                color="red.400"
                fontSize="xs"
                mb={1}
                textAlign="left"
              >
                {error.message}
              </FormHelperText>
            )}
          </Box>
        )}
      />

      <Button
        isDisabled={isDisableSubmit}
        isLoading={isPending}
        colorScheme="brand"
        ml={3}
        aria-label="log-in"
        w="-webkit-fill-available"
        mt={2}
        onClick={handleSubmit(handleLogin)}
      >
        Log In
      </Button>
    </FormControl>
  );
};

export default memo(LoginForm);
