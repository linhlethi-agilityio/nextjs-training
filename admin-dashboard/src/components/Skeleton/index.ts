export * from './SkeletonTableRow';
export { default as PaginationSkeleton } from './SkeletonPagination';
export { default as SkeletonTable } from './SkeletonTable';
export { default as OrderDetailSkeleton } from './OrderDetail';
