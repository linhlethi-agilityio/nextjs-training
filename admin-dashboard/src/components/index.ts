export { default as SearchInput } from './SearchInput';
export { default as SideBarItem } from './SideBarItem';
export { default as SideBar } from './SideBar';
export { default as Pagination } from './Pagination';
export { default as ErrorBoundary } from './ErrorBoundary';
export { default as LoadingIndicator } from './LoadingIndicator';
export { default as ProductActions } from './ProductActions';
export { default as ProductPagination } from './ProductPagination';
export { default as LoginForm } from './LoginForm';
export { default as RegisterForm } from './RegisterForm';
export { default as OrderDetail } from './OrderDetail';
export { default as TableCustomers } from './TableCustomers';
export * from './Dropdown';

// Modals
export * from './Modal/CustomModal';
export { default as ConfirmModal } from './Modal/ConfirmModal';
export { default as OrderModal } from './Modal/OrderModal';

// Skeleton
export * from './Skeleton';

// Table
export { default as TableCell } from './TableCell';
export { default as TableOrder } from './TableOrder';
export * from './TableRow';
export * from './Table';
