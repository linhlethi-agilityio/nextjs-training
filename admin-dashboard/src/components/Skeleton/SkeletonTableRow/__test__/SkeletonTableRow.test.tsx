import { render } from '@testing-library/react';
import { Table, Tbody } from '@chakra-ui/react';

// Components
import { TableColumnType, SkeletonTable } from '@/components';

interface User {
  name: string;
  email: string;
  phone: string;
}

const columnsTable: TableColumnType<User>[] = [
  {
    header: 'Name',
    accessor: 'name',
  },
  {
    header: 'Email',
    accessor: 'email',
  },
  {
    header: 'Phone',
    accessor: 'phone',
  },
];

describe('TableRow', () => {
  it('should match snapshot for TableRow', () => {
    const { container } = render(
      <Table>
        <Tbody>
          <SkeletonTable columns={columnsTable} />
        </Tbody>
      </Table>,
    );

    expect(container).toMatchSnapshot();
  });
});
