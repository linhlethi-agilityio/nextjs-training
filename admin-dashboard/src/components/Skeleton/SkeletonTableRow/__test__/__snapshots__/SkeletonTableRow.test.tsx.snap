// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableRow should match snapshot for TableRow 1`] = `
<div>
  <table
    class="chakra-table css-0"
  >
    <tbody
      class="css-0"
    >
      <tr
        class="css-0"
      >
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
      </tr>
      <tr
        class="css-0"
      >
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
      </tr>
      <tr
        class="css-0"
      >
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
        <td
          class="css-1ys01ni"
        >
          <div
            class="chakra-skeleton css-14aqp8m"
          />
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
